File: src/utils/ner.py
"""Named entity extraction and canonicalization for story grouping."""

import spacy
import re
from typing import List, Dict, Set, Optional, Tuple
from functools import lru_cache
from dataclasses import dataclass

# Load spaCy model (download with: python -m spacy download en_core_web_sm)
@lru_cache
def get_nlp():
    return spacy.load("en_core_web_sm")


ENTITY_LABELS = {"PERSON", "ORG", "GPE", "LOC", "EVENT", "PRODUCT"}


@dataclass
class CanonicalMention:
    """A resolved entity mention with canonical ID."""
    surface_form: str           # What was in the text
    canonical_id: str           # UUID of canonical entity
    canonical_name: str         # Preferred display name
    entity_type: str            # PERSON, ORG, GPE, etc.
    confidence: float = 1.0     # Resolution confidence (0-1)


def extract_entities(text: str, top_n: Optional[int] = 3) -> Dict[str, List[str]]:
    """
    Extract named entities from text.
    Returns dict of label -> list of entity texts (top N by frequency).
    """
    return extract_entities_top_n(text, top_n=top_n)


def extract_entities_top_n(text: str, top_n: Optional[int] = 3) -> Dict[str, List[str]]:
    """
    Extract named entities from text, honoring an explicit top-N cap.

    ``top_n`` is the per-label cap (top N entities per PERSON/ORG/GPE label by
    frequency), so callers can drive it from configuration instead of hardcoding
    the value at each call site. Pass ``top_n=None`` to keep all entities per
    label. ``top_n`` must be >= 0; negative values raise ValueError.
    """
    if top_n is not None and top_n < 0:
        raise ValueError("top_n must be >= 0 (or None to keep all)")
    nlp = get_nlp()
    doc = nlp(text)

    entities_by_label: Dict[str, Dict[str, int]] = {label: {} for label in ENTITY_LABELS}

    for ent in doc.ents:
        if ent.label_ in ENTITY_LABELS:
            entities_by_label[ent.label_][ent.text] = entities_by_label[ent.label_].get(ent.text, 0) + 1

    # Sort by frequency, take top N per label
    result = {}
    for label, counts in entities_by_label.items():
        sorted_entities = sorted(counts.items(), key=lambda x: -x[1])
        if top_n is not None:
            sorted_entities = sorted_entities[:top_n]
        result[label] = [ent for ent, _ in sorted_entities]

    return result


def get_primary_entity_set(entities: Dict[str, List[str]]) -> Set[str]:
    """
    Get a flat set of top entities across PERSON, ORG, GPE for story grouping.
    """
    primary_labels = {"PERSON", "ORG", "GPE"}
    entity_set = set()
    for label in primary_labels:
        entity_set.update(entities.get(label, []))
    return entity_set


def entity_set_jaccard(set_a: Set[str], set_b: Set[str]) -> float:
    """Jaccard similarity between two entity sets."""
    if not set_a and not set_b:
        return 1.0
    if not set_a or not set_b:
        return 0.0
    intersection = set_a & set_b
    union = set_a | set_b
    return len(intersection) / len(union)


# --- Canonicalization layer ---

# Normalization helpers
def _normalize_text(text: str, entity_type: str = "") -> str:
    """Normalize text for matching: lowercase, strip punctuation, collapse whitespace."""
    text = text.lower().strip()
    # Remove common honorifics/titles
    text = re.sub(r'\b(mr\.?|mrs\.?|ms\.?|dr\.?|prof\.?|president|prime minister|pm|secretary|minister)\s+', '', text)
    # Remove punctuation except hyphens in names
    text = re.sub(r'[^\w\s-]', '', text)
    # Collapse whitespace
    text = re.sub(r'\s+', ' ', text)
    normalized = text.strip()
    # Include entity type in normalized key to distinguish same surface form different types
    if entity_type:
        return f"{entity_type}:{normalized}"
    return normalized


def _generate_aliases(base_name: str, entity_type: str) -> List[str]:
    """Generate common aliases for a canonical entity."""
    aliases = set()
    normalized = _normalize_text(base_name)
    aliases.add(normalized)

    # For PERSON: add first/last name variants
    if entity_type == "PERSON":
        parts = normalized.split()
        if len(parts) >= 2:
            # "Joe Biden" -> "biden", "joe biden", "j biden"
            aliases.add(parts[-1])  # Last name only
            aliases.add(f"{parts[0][0]} {parts[-1]}")  # Initial + last
            aliases.add(f"{parts[0]} {parts[-1]}")  # First + last
        if len(parts) >= 3:
            # "President Joe Biden" -> "biden", "joe biden"
            pass

    # For ORG: add acronyms
    if entity_type == "ORG":
        words = normalized.split()
        if len(words) >= 2:
            # "European Union" -> "eu", "e u"
            acronym = ''.join(w[0] for w in words if w)
            if len(acronym) >= 2:
                aliases.add(acronym)
                aliases.add(' '.join(w[0] for w in words))
            # "Federal Bureau of Investigation" -> "fbi"
            # Also add common abbreviation patterns
            for w in words:
                if len(w) > 3:
                    aliases.add(w)

    # For GPE/LOC: add common variants
    if entity_type in ("GPE", "LOC"):
        # "United States" -> "us", "usa", "america"
        if "united states" in normalized:
            aliases.update(["us", "usa", "america"])
        if "united kingdom" in normalized:
            aliases.update(["uk", "britain", "great britain"])
        if "european union" in normalized:
            aliases.update(["eu"])

    return list(aliases)


class EntityCanonicalizer:
    """Resolves entity mentions to canonical entities using database-backed aliases."""

    def __init__(self, session=None):
        self.session = session
        self._cache: Dict[str, CanonicalMention] = {}  # normalized surface -> CanonicalMention
        self._initialized = False

    async def initialize(self):
        """Load all canonical entities and aliases from database."""
        if self._initialized:
            return
        if not self.session:
            return

        from sqlalchemy import select
        from sqlalchemy.orm import selectinload
        from src.schema.models import CanonicalEntity

        # Load all canonical entities with their aliases (eager load to avoid MissingGreenlet)
        stmt = select(CanonicalEntity).options(selectinload(CanonicalEntity.aliases))
        result = await self.session.execute(stmt)
        canonical_entities = result.scalars().all()

        for entity in canonical_entities:
            for alias in entity.aliases:
                self._cache[_normalize_text(alias.alias, entity.entity_type)] = CanonicalMention(
                    surface_form=alias.alias,
                    canonical_id=str(entity.id),
                    canonical_name=entity.canonical_name,
                    entity_type=entity.entity_type,
                    confidence=1.0
                )
            # Also cache the canonical name itself
            self._cache[_normalize_text(entity.canonical_name, entity.entity_type)] = CanonicalMention(
                surface_form=entity.canonical_name,
                canonical_id=str(entity.id),
                canonical_name=entity.canonical_name,
                entity_type=entity.entity_type,
                confidence=1.0
            )

        self._initialized = True

    def resolve(self, surface_form: str, entity_type: str) -> Optional[CanonicalMention]:
        """
        Resolve a surface form to a canonical entity.
        Returns None if no match found (caller should create new canonical entity).
        """
        if not self._initialized:
            return None

        normalized = _normalize_text(surface_form, entity_type)
        if normalized in self._cache:
            return self._cache[normalized]

        # Try fuzzy matching for common variations
        for cached_key, mention in self._cache.items():
            if mention.entity_type != entity_type:
                continue
            # Simple fuzzy: check if one contains the other
            if normalized in cached_key or cached_key in normalized:
                # Additional check: they should share significant tokens
                norm_tokens = set(normalized.split())
                cached_tokens = set(cached_key.split())
                if norm_tokens & cached_tokens:
                    return CanonicalMention(
                        surface_form=surface_form,
                        canonical_id=mention.canonical_id,
                        canonical_name=mention.canonical_name,
                        entity_type=entity_type,
                        confidence=0.8
                    )

        return None

    async def get_or_create(self, surface_form: str, entity_type: str) -> CanonicalMention:
        """
        Resolve or create a canonical entity for a surface form.
        Creates new canonical entity + aliases if not found.
        """
        # Try to resolve first
        resolved = self.resolve(surface_form, entity_type)
        if resolved:
            return resolved

        # Not found - create new canonical entity
        if not self.session:
            # Fallback: return unresolved mention with generated ID
            import uuid
            return CanonicalMention(
                surface_form=surface_form,
                canonical_id=str(uuid.uuid4()),
                canonical_name=surface_form,
                entity_type=entity_type,
                confidence=0.5
            )

        from src.schema.models import CanonicalEntity, EntityAlias
        import uuid

        # Create canonical entity
        canonical_id = uuid.uuid4()
        canonical_entity = CanonicalEntity(
            id=canonical_id,
            canonical_name=surface_form,
            entity_type=entity_type
        )
        self.session.add(canonical_entity)

        # Create initial alias (the surface form itself)
        alias = EntityAlias(
            canonical_entity_id=canonical_id,
            alias=surface_form
        )
        self.session.add(alias)

        # Generate and add common aliases
        for alias_text in _generate_aliases(surface_form, entity_type):
            normalized_alias = _normalize_text(alias_text, entity_type)
            if normalized_alias != _normalize_text(surface_form, entity_type):
                alias_obj = EntityAlias(
                    canonical_entity_id=canonical_id,
                    alias=alias_text
                )
                self.session.add(alias_obj)
                # Add to cache
                self._cache[normalized_alias] = CanonicalMention(
                    surface_form=alias_text,
                    canonical_id=str(canonical_id),
                    canonical_name=surface_form,
                    entity_type=entity_type,
                    confidence=0.9
                )

        await self.session.flush()

        # Add the main entry to cache
        normalized = _normalize_text(surface_form, entity_type)
        mention = CanonicalMention(
            surface_form=surface_form,
            canonical_id=str(canonical_id),
            canonical_name=surface_form,
            entity_type=entity_type,
            confidence=1.0
        )
        self._cache[normalized] = mention

        return mention


async def resolve_entities_to_canonical(
    entities: Dict[str, List[str]],
    canonicalizer: EntityCanonicalizer
) -> Tuple[Set[str], List[CanonicalMention]]:
    """
    Resolve extracted entities to canonical IDs.
    Returns (canonical_id_set, canonical_mentions) for story grouping.
    """
    canonical_ids = set()
    mentions = []

    for entity_type, surface_forms in entities.items():
        for surface in surface_forms:
            mention = await canonicalizer.get_or_create(surface, entity_type)
            canonical_ids.add(mention.canonical_id)
            mentions.append(mention)

    return canonical_ids, mentions


def canonical_jaccard(set_a: Set[str], set_b: Set[str]) -> float:
    """Jaccard similarity on canonical entity ID sets."""
    if not set_a and not set_b:
        return 1.0
    if not set_a or not set_b:
        return 0.0
    intersection = set_a & set_b
    union = set_a | set_b
    return len(intersection) / len(union)

File: pipeline-rs/src/utils/ner.rs
//! Entity canonicalization for story grouping
//!
//! This module handles everything *except* the spaCy NER call itself.
//! The actual NER extraction (spaCy -> entities) is in T5.
//! Here we implement: normalization, alias generation, EntityCanonicalizer,
//! resolve_entities_to_canonical, canonical_jaccard.

use crate::models::{CanonicalEntity, EntityAlias};
use crate::database::PgPool;
use regex::Regex;
use serde::{Deserialize, Serialize};
use std::collections::{HashMap, HashSet};
use std::sync::Arc;
use tokio::sync::RwLock;
use tracing::{debug, warn};
use uuid::Uuid;

/// Entity labels we care about (OntoNotes style)
pub const ENTITY_LABELS: &[&str] = &["PERSON", "ORG", "GPE", "LOC", "EVENT", "PRODUCT"];

/// Primary entity labels for story grouping
pub const PRIMARY_ENTITY_LABELS: &[&str] = &["PERSON", "ORG", "GPE"];

/// A resolved entity mention with canonical ID
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CanonicalMention {
    pub surface_form: String,   // What was in the text
    pub canonical_id: String,   // UUID of canonical entity
    pub canonical_name: String, // Preferred display name
    pub entity_type: String,    // PERSON, ORG, GPE, etc.
    pub confidence: f64,        // Resolution confidence (0-1)
}

/// Extracted entities format: Dict<label, Vec<String>>
pub type EntitiesDict = HashMap<String, Vec<String>>;

/// EntityCanonicalizer resolves entity mentions to canonical entities using database-backed aliases
pub struct EntityCanonicalizer {
    pool: Arc<PgPool>,
    cache: Arc<RwLock<HashMap<String, CanonicalMention>>>, // normalized surface -> CanonicalMention
    initialized: bool,
}

impl EntityCanonicalizer {
    pub fn new(pool: Arc<PgPool>) -> Self {
        Self {
            pool,
            cache: Arc::new(RwLock::new(HashMap::new())),
            initialized: false,
        }
    }

    /// Check if initialized
    pub fn is_initialized(&self) -> bool {
        self.initialized
    }

    /// Load all canonical entities and aliases from database
    pub async fn initialize(&mut self) -> Result<(), sqlx::Error> {
        if self.initialized {
            return Ok(());
        }

        // Load all canonical entities with their aliases
        let rows = sqlx::query!(
            r#"
            SELECT ce.id, ce.canonical_name, ce.entity_type, ea.alias
            FROM canonical_entities ce
            LEFT JOIN entity_aliases ea ON ea.canonical_entity_id = ce.id
            "#
        )
        .fetch_all(&*self.pool)
        .await?;

        let mut cache = self.cache.write().await;
        for row in rows {
            let canonical_id = row.id.to_string();
            let canonical_name = row.canonical_name;
            let entity_type = row.entity_type;

            // Add alias if present
            if let Some(alias) = row.alias {
                let normalized = Self::normalize_text(&alias, &entity_type);
                cache.insert(
                    normalized,
                    CanonicalMention {
                        surface_form: alias,
                        canonical_id: canonical_id.clone(),
                        canonical_name: canonical_name.clone(),
                        entity_type: entity_type.clone(),
                        confidence: 1.0,
                    },
                );
            }

            // Also cache the canonical name itself
            let normalized = Self::normalize_text(&canonical_name, &entity_type);
            cache.insert(
                normalized,
                CanonicalMention {
                    surface_form: canonical_name.clone(),
                    canonical_id: canonical_id.clone(),
                    canonical_name,
                    entity_type: entity_type.clone(),
                    confidence: 1.0,
                },
            );
        }

        self.initialized = true;
        debug!("EntityCanonicalizer initialized with {} entries", cache.len());
        Ok(())
    }

    /// Normalize text for matching: lowercase, strip punctuation, collapse whitespace
    pub fn normalize_text(text: &str, entity_type: &str) -> String {
        let mut text = text.to_lowercase().trim().to_string();

        // Remove common honorifics/titles
        let honorifics = Regex::new(r"\b(mr\.?|mrs\.?|ms\.?|dr\.?|prof\.?|president|prime minister|pm|secretary|minister)\s+").unwrap();
        text = honorifics.replace_all(&text, "").to_string();

        // Remove punctuation except hyphens in names
        let punct = Regex::new(r"[^\w\s-]").unwrap();
        text = punct.replace_all(&text, "").to_string();

        // Collapse whitespace
        let ws = Regex::new(r"\s+").unwrap();
        text = ws.replace_all(&text, " ").to_string();

        let normalized = text.trim().to_string();

        // Include entity type in normalized key to distinguish same surface form different types
        if !entity_type.is_empty() {
            format!("{}:{}", entity_type, normalized)
        } else {
            normalized
        }
    }

    /// Generate common aliases for a canonical entity
    pub fn generate_aliases(base_name: &str, entity_type: &str) -> Vec<String> {
        let mut aliases = HashSet::new();
        let normalized = Self::normalize_text(base_name, "");
        aliases.insert(normalized);

        match entity_type {
            "PERSON" => {
                let parts: Vec<&str> = normalized.split_whitespace().collect();
                if parts.len() >= 2 {
                    // "joe biden" -> "biden", "j biden"
                    aliases.insert(parts.last().unwrap().to_string());
                    aliases.insert(format!("{} {}", parts[0].chars().next().unwrap(), parts.last().unwrap()));
                    aliases.insert(format!("{} {}", parts[0], parts.last().unwrap()));
                }
            }
            "ORG" => {
                let words: Vec<&str> = normalized.split_whitespace().collect();
                if words.len() >= 2 {
                    // "european union" -> "eu", "e u"
                    let acronym: String = words.iter().filter(|w| !w.is_empty()).map(|w| w.chars().next().unwrap()).collect();
                    if acronym.len() >= 2 {
                        aliases.insert(acronym.clone());
                        aliases.insert(words.iter().map(|w| w.chars().next().unwrap()).collect::<String>());
                    }
                    // Also add individual words > 3 chars
                    for w in words {
                        if w.len() > 3 {
                            aliases.insert(w.to_string());
                        }
                    }
                }
            }
            "GPE" | "LOC" => {
                // Common country/region variants
                if normalized.contains("united states") {
                    aliases.extend(["us".to_string(), "usa".to_string(), "america".to_string()]);
                }
                if normalized.contains("united kingdom") {
                    aliases.extend(["uk".to_string(), "britain".to_string(), "great britain".to_string()]);
                }
                if normalized.contains("european union") {
                    aliases.insert("eu".to_string());
                }
            }
            _ => {}
        }

        aliases.into_iter().collect()
    }

    /// Resolve a surface form to a canonical entity
    /// Returns None if no match found (caller should create new canonical entity)
    pub async fn resolve(&self, surface_form: &str, entity_type: &str) -> Option<CanonicalMention> {
        if !self.initialized {
            return None;
        }

        let normalized = Self::normalize_text(surface_form, entity_type);
        {
            let cache = self.cache.read().await;
            if let Some(mention) = cache.get(&normalized) {
                return Some(mention.clone());
            }
        }

        // Try fuzzy matching for common variations
        let cache = self.cache.read().await;
        for (cached_key, mention) in cache.iter() {
            if mention.entity_type != entity_type {
                continue;
            }
            // Simple fuzzy: check if one contains the other
            if normalized.contains(cached_key.as_str()) || cached_key.contains(normalized.as_str()) {
                // Additional check: they should share significant tokens
                let norm_tokens: HashSet<&str> = normalized.split_whitespace().collect();
                let cached_tokens: HashSet<&str> = cached_key.split_whitespace().collect();
                if !norm_tokens.is_disjoint(&cached_tokens) {
                    return Some(CanonicalMention {
                        surface_form: surface_form.to_string(),
                        canonical_id: mention.canonical_id.clone(),
                        canonical_name: mention.canonical_name.clone(),
                        entity_type: entity_type.to_string(),
                        confidence: 0.8,
                    });
                }
            }
        }

        None
    }

    /// Resolve or create a canonical entity for a surface form
    /// Creates new canonical entity + aliases if not found
    pub async fn get_or_create(&self, surface_form: &str, entity_type: &str) -> Result<CanonicalMention, sqlx::Error> {
        // Try to resolve first
        if let Some(resolved) = self.resolve(surface_form, entity_type).await {
            return Ok(resolved);
        }

        // Not found - create new canonical entity
        let canonical_id = Uuid::new_v4();
        let canonical_id_str = canonical_id.to_string();

        // Create canonical entity
        sqlx::query!(
            r#"
            INSERT INTO canonical_entities (id, canonical_name, entity_type)
            VALUES ($1, $2, $3)
            "#,
            canonical_id,
            surface_form,
            entity_type
        )
        .execute(&*self.pool)
        .await?;

        // Create initial alias (the surface form itself)
        sqlx::query!(
            r#"
            INSERT INTO entity_aliases (canonical_entity_id, alias)
            VALUES ($1, $2)
            ON CONFLICT (canonical_entity_id, alias) DO NOTHING
            "#,
            canonical_id,
            surface_form
        )
        .execute(&*self.pool)
        .await?;

        // Generate and add common aliases
        for alias_text in Self::generate_aliases(surface_form, entity_type) {
            let normalized_alias = Self::normalize_text(&alias_text, entity_type);
            let normalized_main = Self::normalize_text(surface_form, entity_type);
            if normalized_alias != normalized_main {
                sqlx::query!(
                    r#"
                    INSERT INTO entity_aliases (canonical_entity_id, alias)
                    VALUES ($1, $2)
                    ON CONFLICT (canonical_entity_id, alias) DO NOTHING
                    "#,
                    canonical_id,
                    alias_text
                )
                .execute(&*self.pool)
                .await?;

                // Add to cache
                let mut cache = self.cache.write().await;
                cache.insert(
                    normalized_alias,
                    CanonicalMention {
                        surface_form: alias_text,
                        canonical_id: canonical_id_str.clone(),
                        canonical_name: surface_form.to_string(),
                        entity_type: entity_type.to_string(),
                        confidence: 0.9,
                    },
                );
            }
        }

        // Add the main entry to cache
        let normalized = Self::normalize_text(surface_form, entity_type);
        let mention = CanonicalMention {
            surface_form: surface_form.to_string(),
            canonical_id: canonical_id_str,
            canonical_name: surface_form.to_string(),
            entity_type: entity_type.to_string(),
            confidence: 1.0,
        };

        let mut cache = self.cache.write().await;
        cache.insert(normalized, mention.clone());

        debug!("Created new canonical entity: {} ({})", surface_form, entity_type);
        Ok(mention)
    }
}

/// Resolve extracted entities to canonical IDs
/// Returns (canonical_id_set, canonical_mentions) for story grouping
pub async fn resolve_entities_to_canonical(
    entities: &EntitiesDict,
    canonicalizer: &EntityCanonicalizer,
) -> Result<(HashSet<String>, Vec<CanonicalMention>), sqlx::Error> {
    let mut canonical_ids = HashSet::new();
    let mut mentions = Vec::new();

    for (entity_type, surface_forms) in entities {
        for surface in surface_forms {
            let mention = canonicalizer.get_or_create(surface, entity_type).await?;
            canonical_ids.insert(mention.canonical_id.clone());
            mentions.push(mention);
        }
    }

    Ok((canonical_ids, mentions))
}

/// Get a flat set of top entities across PERSON, ORG, GPE for story grouping
pub fn get_primary_entity_set(entities: &EntitiesDict) -> HashSet<String> {
    let mut entity_set = HashSet::new();
    for label in PRIMARY_ENTITY_LABELS {
        if let Some(entities_for_label) = entities.get(*label) {
            entity_set.extend(entities_for_label.iter().cloned());
        }
    }
    entity_set
}

/// Jaccard similarity between two entity sets
pub fn entity_set_jaccard(set_a: &HashSet<String>, set_b: &HashSet<String>) -> f64 {
    if set_a.is_empty() && set_b.is_empty() {
        return 1.0;
    }
    if set_a.is_empty() || set_b.is_empty() {
        return 0.0;
    }
    let intersection = set_a.intersection(set_b).count();
    let union = set_a.union(set_b).count();
    intersection as f64 / union as f64
}

/// Jaccard similarity on canonical entity ID sets
pub fn canonical_jaccard(set_a: &HashSet<String>, set_b: &HashSet<String>) -> f64 {
    if set_a.is_empty() && set_b.is_empty() {
        return 1.0;
    }
    if set_a.is_empty() || set_b.is_empty() {
        return 0.0;
    }
    let intersection = set_a.intersection(set_b).count();
    let union = set_a.union(set_b).count();
    intersection as f64 / union as f64
}

/// Extract entities from text - PLACEHOLDER for T5
/// In T5, this will call rust-bert or onnxruntime for actual NER
pub async fn extract_entities_top_n(_text: &str, _top_n: Option<usize>) -> EntitiesDict {
    // Placeholder - returns empty dict
    // In T5, this will be implemented with rust-bert/onnxruntime
    EntitiesDict::new()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_normalize_text() {
        let normalized = EntityCanonicalizer::normalize_text("  Mr. Joe Biden  ", "PERSON");
        assert_eq!(normalized, "person:joe biden");
    }

    #[test]
    fn test_normalize_text_org() {
        let normalized = EntityCanonicalizer::normalize_text("European Union", "ORG");
        assert_eq!(normalized, "org:european union");
    }

    #[test]
    fn test_generate_aliases_person() {
        let aliases = EntityCanonicalizer::generate_aliases("Joe Biden", "PERSON");
        assert!(aliases.contains(&"biden".to_string()));
        assert!(aliases.contains(&"j biden".to_string()));
    }

    #[test]
    fn test_generate_aliases_org() {
        let aliases = EntityCanonicalizer::generate_aliases("European Union", "ORG");
        assert!(aliases.contains(&"eu".to_string()));
    }

    #[test]
    fn test_canonical_jaccard() {
        let mut set_a = HashSet::new();
        set_a.insert("1".to_string());
        set_a.insert("2".to_string());

        let mut set_b = HashSet::new();
        set_b.insert("2".to_string());
        set_b.insert("3".to_string());

        let j = canonical_jaccard(&set_a, &set_b);
        assert!((j - 1.0 / 3.0).abs() < 0.001);
    }

    #[test]
    fn test_entity_set_jaccard() {
        let mut set_a = HashSet::new();
        set_a.insert("Biden".to_string());
        set_a.insert("White House".to_string());

        let mut set_b = HashSet::new();
        set_b.insert("Biden".to_string());
        set_b.insert("Oval Office".to_string());

        let j = entity_set_jaccard(&set_a, &set_b);
        assert!((j - 1.0 / 3.0).abs() < 0.001);
    }
}